[core][compiled graph]Skip python-based deserialization upon CompiledDAGRef's destruction

jeffreyjeffreywang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
Currently, if a CompiledDAGRef is not retrieved with `get()`, it releases its native buffer and deserializes the result upon its destruction. The deserialized result is not used anywhere, and we should skip the deserialization step and only release the underlying native buffer for future data.

## Related issue number

Resolves #47614 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

kevin85421 commented on python/ray/experimental/channel/shared_memory_channel.py:
```diff
@@ -750,6 +777,10 @@ def read(self, timeout: Optional[float] = None) -> Any:
 
+    def release_buffer(self, timeout: Optional[float] = None):
+        self.ensure_registered_as_reader()
+        self._channel_dict[self._get_self_actor_id()].release_buffer(timeout)
```
Is it possible that the channel is an IntraProcessChannel? I think it's unlikely, but I just want to double-check.

jeffreyjeffreywang replied:
Afaik, communications between actors are conducted through `IntraProcessChannel`, and communications between the driver and actor go through `BufferedSharedMemoryChannel`. Destruction of `CompiledDAGRef` happens on the driver, and the deserialization I skipped occurs between the driver and an actor. Therefore, the underlying channel is always `BufferedSharedMemoryChannel`.

I stepped into some unit tests where `IntraProcessChannel` is used (e.g. `TestCompositeChannel::test_composite_channel_one_actor`) and verified that DAG only uses `BufferedSharedMemoryChannel` to write its outputs.

Let me know if I misunderstood anything! I'm still in the process of getting more familiar with the Ray backend :)

kevin85421 replied:
It is correct that the channels between the driver process and actors must be shared memory channels.

kevin85421 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -174,6 +174,29 @@ def test_basic(ray_start_regular):
+            "The native buffer associated with the CompiledDAGRef was not "
+            "released upon destruction."
+        )
+
```
can you call `execute` again and make sure `ray.get` can get the correct result?

jeffreyjeffreywang replied:
Yep, added below!

jeffreyjeffreywang replied:
Thanks for the review btw!

kevin85421 reviewed (commented):
Overall LGTM!

jcotant1 commented:
Hey @jeffreyjeffreywang @kevin85421 what's the status of this one?

kevin85421 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -2056,6 +2056,21 @@ def _get_execution_results(
                 del self._result_buffer[execution_index]
         return result
 
+    def _release_output_channel_buffers(self, execution_index: int):
```
Typically, if a class method starts with `_` in the Ray codebase, it is expected to be called only by other methods within the same class. How about removing `_` if it is called in `compiled_dag_ref.py`?

jeffreyjeffreywang commented:
Sorry for the delayed response. I've addressed your comment @kevin85421! Thanks again for reviewing my PR. Could you please help me merge the PR once approved? I don't think I have the permissions to merge PRs.

kevin85421 commented:
@jeffreyjeffreywang CI tests failed. Would you mind taking a look?

jeffreyjeffreywang commented:
I took a closer look at the [failures](https://buildkite.com/ray-project/premerge/builds/31575#01938af4-d831-4328-85a7-21704359e707). Most of them are not relevant to the changes in this PR. For those which are relevant (e.g. `test_mocked_nccl_dag.py` and `test_multi_node_dag.py`), they all succeeded locally. I'm not sure where the errors resulted from when inspecting the failure logs. I'll rebase and kick off another run to see if errors reproduce.

Louis-Groux commented on python/ray/experimental/channel/shared_memory_channel.py:
```diff
@@ -503,6 +503,18 @@ def read(self, timeout: Optional[float] = None) -> Any:
         return ret
 
+    def release_buffer(self, timeout: Optional[float] = None) -> None:
+        assert (
+            timeout is None or timeout >= 0 or timeout == -1
+        ), "Timeout must be non-negative or -1."
```
Should this be moved closer to where the timeout is acquired? In `compiled_dag_node.py:release_output_channel_buffers` perhaps?

Louis-Groux commented on python/ray/experimental/channel/shared_memory_channel.py:
```diff
@@ -604,6 +616,21 @@ def read(self, timeout: Optional[float] = None) -> Any:
+        # A single channel is not supposed to read and write at the same time.
+        assert self._next_write_index == 0
+        self._buffers[self._next_read_index].release_buffer(timeout)
+        self._next_read_index += 1
+        self._next_read_index %= self._num_shm_buffers
```
Nit, you can one-line this if ray is okay with ~85 chars long lines:
```suggestion
        self._next_read_index = (self._next_read_index + 1) % self._num_shm_buffers
```

jeffreyjeffreywang commented:
Different set of tests failed, including the test I introduced in the PR. I'll take a look at the failure.

jeffreyjeffreywang commented:
I needed to adjust how `actor_id` is determined when releasing the underlying buffer, as reading from the driver is now supported as of https://github.com/ray-project/ray/pull/48963. The failed tests now passed locally. @kevin85421 would you mind taking a look at the latest update once you have time? Thank you :)

jeffreyjeffreywang commented:
I don't think the remaining test failures are related to this PR. They result from serve's doc gpu tests.

kevin85421 commented:
I rerun the test

jeffreyjeffreywang commented:
Looks like `//doc:source/serve/doc_code/object_detection` consistently fails with the following error message. Will look into this particular test to see if it has any dependency on what I changed in my PR. At first glance, `get_objects` shows up in the trace. I introduced a new argument `skip_deserialization` with default value `False` to the `get_objects` API. It shouldn't affect existing callers.

```
[2024-12-07T07:12:48Z] (ServeController pid=1477) Traceback (most recent call last):
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/serve/_private/deployment_state.py", line 696, in check_ready
[2024-12-07T07:12:48Z] (ServeController pid=1477)     ) = ray.get(self._ready_obj_ref)
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
[2024-12-07T07:12:48Z] (ServeController pid=1477)     return fn(*args, **kwargs)
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/_private/client_mode_hook.py", line 103, in wrapper
[2024-12-07T07:12:48Z] (ServeController pid=1477)     return func(*args, **kwargs)
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/_private/worker.py", line 2773, in get
[2024-12-07T07:12:48Z] (ServeController pid=1477)     values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/_private/worker.py", line 920, in get_objects
[2024-12-07T07:12:48Z] (ServeController pid=1477)     raise value.as_instanceof_cause()
[2024-12-07T07:12:48Z] (ServeController pid=1477) ray.exceptions.RayTaskError(RuntimeError): ray::ServeReplica:default:ObjectDetection.initialize_and_get_metadata() (pid=1563, ip=172.16.0.3, actor_id=7c4dae6709dd6f417b03f63701000000, repr=<ray.serve._private.replica.ServeReplica:default:ObjectDetection object at 0x7f00c51dcca0>)
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/opt/miniconda/lib/python3.9/concurrent/futures/_base.py", line 446, in result
[2024-12-07T07:12:48Z] (ServeController pid=1477)     return self.__get_result()
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/opt/miniconda/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
[2024-12-07T07:12:48Z] (ServeController pid=1477)     raise self._exception
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/serve/_private/replica.py", line 887, in initialize_and_get_metadata
[2024-12-07T07:12:48Z] (ServeController pid=1477)     await self._replica_impl.initialize(deployment_config)
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/rayci/python/ray/serve/_private/replica.py", line 626, in initialize
[2024-12-07T07:12:48Z] (ServeController pid=1477)     raise RuntimeError(traceback.format_exc()) from None
[2024-12-07T07:12:48Z] (ServeController pid=1477) RuntimeError: Traceback (most recent call last):
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/root/.cache/torch/hub/ultralytics_yolov5_master/models/common.py", line 27, in <module>
[2024-12-07T07:12:48Z] (ServeController pid=1477)     import ultralytics
[2024-12-07T07:12:48Z] (ServeController pid=1477)   File "/tmp/ray/session_2024-12-07_07-10-04_055869_1087/runtime_resources/pip/9eb89356eed531a9f1e3204027511a50551c183e/virtualenv/lib/python3.9/site-packages/ultralytics/__init__.py", line 6, in <module>
[2024-12-07T07:12:48Z] (ServeController pid=1477)     import package
[2024-12-07T07:12:48Z] (ServeController pid=1477) ModuleNotFoundError: No module named 'package'
```

kevin85421 commented:
Rerun the test again. If it still fails, you may need to dive into the error.

jeffreyjeffreywang commented:
Thanks Kevin. The latest rerun appears to have a different error: requested docker image wasn't found in ECR.
```
[2024-12-09T19:27:43Z] Unable to find image '029272617770.dkr.ecr.us-west-2.amazonaws.com/rayproject/citemp:d8c902d0-forge' locally
[2024-12-09T19:27:43Z] docker: Error response from daemon: manifest for 029272617770.dkr.ecr.us-west-2.amazonaws.com/rayproject/citemp:d8c902d0-forge not found: manifest unknown: Requested image not found.
```

I'm looking into what the root cause is.

jeffreyjeffreywang commented:
Looks like the former errors are transient. All tests have passed. @kevin85421 would you mind merging this PR once you have time? Thank you 😃

kevin85421 commented:
Nice! I don't have the access to merge PR, but I will ping my colleague to merge it.
